crates/scheduler/src/batch_scheduler.rs

```diff
@@ -0,0 +1,448 @@
+                        update_cap,
+                    );
+                    let peer_position = tracker.worker_tracker.worker_position(&peer_id)?;
+                    tracing::debug!(time = %time, count = %cnt, "Finish simulation");
+                    tracing::info!(
+                        "Simulation with projection {:?} and {:?} of {:?}",
+                        projection,
+                        cnt,
+                        tracker.count()
+                    );
+                    if cnt == 0 && projection[peer_position] < update_cap {
+                        tracker
+                            .worker_tracker
+                            .update_worker_state(&peer_id, State::UpdateScheduled);
+                        Ok(progress::Response::ScheduleUpdate {
+                            counter: projection[peer_position],
+                        })
+                    } else {
+                        Ok(progress::Response::Continue {})
+                    }
+                }
+                State::UpdateScheduled => Ok(progress::Response::Continue {}),
+                _ => Err(BatchSchedulerError::Scheduling("Wrong state".into())),
+            }
+        }
+        progress::Progress::Update => {
+            tracker
+                .lock()
+                .await
+                .worker_tracker
+                .update_worker_state(&peer_id, State::Updating);
+            Ok(progress::Response::Ok {})
+        }
+        progress::Progress::Updated => {
+            let mut tracker = tracker.lock().await;
+            tracker.next_round();
+            if tracker.training_finished() {
+                Ok(progress::Response::Done {})
+            } else {
+                Ok(progress::Response::Ok {})
+            }
+        }
+        progress::Progress::UpdateReceived => {
+            let mut tracker = tracker.lock().await;
+            if tracker.training_finished() {
+                tracker
+                    .worker_tracker
+                    .update_worker_state(&peer_id, State::Done);
+                if tracker
+                    .worker_tracker
+                    .states()
+                    .iter()
+                    .all(|state| state == &State::Done)
+                {
+                    notifier.notify_one();
+                }
+                Ok(progress::Response::Done {})
+            } else {
+                tracker
+                    .worker_tracker
+                    .update_worker_state(&peer_id, State::Training);
+                Ok(progress::Response::Continue {})
+            }
+        }
+    }
+}
+
+pub struct BatchScheduler {}
+
+impl BatchScheduler {
+    pub async fn run<T>(
+        network: Network,
+        simulator: SimulatorFn,
+        tracker: Arc<Mutex<ProgressTracker<T>>>,
+        id: Uuid,
+    ) -> Result<(mpsc::Receiver<(PeerId, Metrics)>, JoinHandle<()>), BatchSchedulerError>
+    where
+        T: RuntimeStatistic + 'static,
+    {
```

Comment:
Instead of passing the `SimulatorFn` let's select the `Simulation` like the `RuntimeStatistic`. 

```suggestion
    pub async fn run<T, S>(
        network: Network,
        tracker: Arc<Mutex<ProgressTracker<T>>>,
        id: Uuid,
    ) -> Result<(mpsc::Receiver<(PeerId, Metrics)>, JoinHandle<()>), BatchSchedulerError>
    where
        T: RuntimeStatistic + 'static,
        S: Simulation + Send + Sync + 'static,
    {
```

And pass it the same way to the the `handle_request` and use it like this:

```
 S::project(
        tracker.worker_tracker.last_updates(),
        tracker.worker_tracker.batch_sizes(),
        tracker.worker_tracker.estimates(),
        tracker.count(),
        time_cap,
        update_cap,
    );
```

With that change we can get rid of the _duplicated_ `SimulatorFn`.